src/append_only_data.rs

```diff
@@ -429,10 +432,46 @@ macro_rules! impl_appendable_data {
                 self.inner.permissions.len() as u64
             }
 
-            // fn user_permissions(&self, user: &User) -> Result<&PubPermissionSet> {
-            //     let perm_set = &self.inner.permissions[self.inner.permissions.len() - 1];
-            //     perm_set.permissions.get(user).ok_or(Error::X)
-            // }
+            fn check_permissions_for_key(
+                &self,
+                requester: PublicKey,
+                permissions: &P,
+                request: Request,
+            ) -> Result<()> {
+                match request {
+                    Request::GetAData(..)
+                    | Request::GetADataShell { .. }
+                    | Request::GetADataRange { .. }
+                    | Request::GetADataIndices(..)
+                    | Request::GetADataLastEntry(..)
+                    | Request::GetADataPermissions { .. }
+                    | Request::GetPubADataUserPermissions { .. }
+                    | Request::GetUnpubADataUserPermissions { .. }
+                    | Request::GetADataOwners { .. } => {
+                        if permissions.is_action_allowed(requester, Action::Read) {
+                            Ok(())
+                        } else {
+                            Err(Error::AccessDenied)
+                        }
+                    }
+                    Request::AddPubADataPermissions { .. }
+                    | Request::AddUnpubADataPermissions { .. }
+                    | Request::SetADataOwner { .. } => {
```

Comment:
`Request::SetADataOwner` is available only to an existing owner, so it should be always `Error::AccessDenied` for non-owner requests.